chart/assets/operations/instance_groups/api.yaml

```diff
@@ -32,6 +32,27 @@
   path: /instance_groups/name=api/jobs/name=cloud_controller_ng/properties/ccdb/connection_validation_timeout?
   value: 3600
 
+# Default settings for the CF info endpoints.
+##
+# Note that the version information has to be a plain integer, as per the
+# Membrane::SchemaValidation occuring in `bosh-pre-start-cloud-controller-ng`.
+
+- type: replace
+  path: /instance_groups/name=api/jobs/name=cloud_controller_ng/properties/name?
+  value: 'KubeCF'
+- type: replace
+  path: /instance_groups/name=api/jobs/name=cloud_controller_ng/properties/build?
+  value: {{ default .Chart.Version .Chart.AppVersion | quote }}
+- type: replace
+  path: /instance_groups/name=api/jobs/name=cloud_controller_ng/properties/version?
+  value: 2
+- type: replace
+  path: /instance_groups/name=api/jobs/name=cloud_controller_ng/properties/support_address?
+  value: 'https:// ??? FIXME ???'
+- type: replace
+  path: /instance_groups/name=api/jobs/name=cloud_controller_ng/properties/description?
+  value: 'CloudFoundry KubeCF'
+
```

Comment:
Instead of a series of ops, I would write this in a config file, e.g. `config/api.yaml`:

```
properties:
  api:
    cloud_controller_ng:
      name: KubeCF
      version: 2
```

Of course this doesn't work for `build` because we currently don't support template expressions in the config files (but that might be fixable). Just leave a comment in the config file that explains that the `build` value is set via ops.

Reply:
+1. Should make it easier to modify when KubeCF is deployed as part of CAP.

Reply:
> we currently don't support template expressions in the config files (but that might be fixable).

I would try this (in https://github.com/cloudfoundry-incubator/kubecf/blob/master/chart/templates/_config.tpl#L49)

change

```
{{- $_ := set $configs $name ($bytes | toString | fromYaml) }}
```

to

```
{{- $_ := set $configs $name ($bytes | toString | tpl | fromYaml) }}
```

I think it should work will become generally useful.

Reply:
Ok. Will try that in a second iteration of my current changes. Which implement the basic move to config/api.yml.

Reply:
> Should make it easier to modify when KubeCF is deployed as part of CAP.

It doesn't matter; CAP could still use `properties` to overwrite whatever you are generating with your ops. Internally `properties` still generates the same kind of ops, but they are applied last, so they can overwrite *everything*: https://github.com/cloudfoundry-incubator/kubecf/blob/master/chart/templates/bosh_deployment.yaml#L90-L91

I still prefer to move config stuff as much as possible out of ops files into simple config data because it is easier to maintain.

Reply:
If you add the `tpl` functionality, please also add a comment that explains that templates in config files only have access to global helm variables, plus the user-provided values from `values.yaml`. They are evaluated *before* all the config files are merged into `$.Values`, so none of the internal config data can be used.

Reply:
> none of the internal config data can be used

Otherwise config files would become self-referential. We won't allow that to make sure our helm charts don't become self-aware. Given the complexities of the templates this could be a real danger... 😛

Reply:
@jandubois note that the insertion of `tpl` into the pipeline does not work:
```
executing "_config.load" at <tpl>: wrong number of args for tpl: want 2 got 1
```

Reply:
It requires a path as first argument ? The name of the file we are processing ?

Reply:
> It requires a path as first argument ? The name of the file we are processing ?

No, it requires the context as the second argument. Try this:

```
{{- $_ := set $configs $name (tpl ($bytes | toString) $) | fromYaml) }}
```

Reply:
After reading docs the proper incantation looks to be
```
{{- $_ := set $configs $name (tpl ($bytes | toString) $ | fromYaml)}}
```

Reply:
Done with commit ccd904248b11b735a4df5c43391c5984a486119f